'use strict';

var util = require('util');

const mongoErrorContextSymbol = Symbol('mongoErrorContextSymbol');

/**
 * Creates a new MongoError
 * @class
 * @augments Error
 * @param ***REMOVED***Error|string|object***REMOVED*** message The error message
 * @property ***REMOVED***string***REMOVED*** message The error message
 * @property ***REMOVED***string***REMOVED*** stack The error call stack
 * @return ***REMOVED***MongoError***REMOVED*** A MongoError instance
 */
function MongoError(message) ***REMOVED***
  var tmp = Error.apply(this, arguments);
  tmp.name = this.name = 'MongoError';

  if (message instanceof Error) ***REMOVED***
    this.message = message.message;
    this.stack = message.stack;
  ***REMOVED*** else ***REMOVED***
    if (typeof message === 'string') ***REMOVED***
      this.message = message;
    ***REMOVED*** else ***REMOVED***
      this.message = message.message || message.errmsg || message.$err || 'n/a';
      for (var name in message) ***REMOVED***
        this[name] = message[name];
      ***REMOVED***
    ***REMOVED***
    if (Error.captureStackTrace) ***REMOVED***
      Error.captureStackTrace(this, this.constructor);
    ***REMOVED***
  ***REMOVED***

  this[mongoErrorContextSymbol] = this[mongoErrorContextSymbol] || ***REMOVED******REMOVED***;
***REMOVED***
util.inherits(MongoError, Error);

/**
 * Creates a new MongoError object
 * @method
 * @param ***REMOVED***Error|string|object***REMOVED*** options The options used to create the error.
 * @return ***REMOVED***MongoError***REMOVED*** A MongoError instance
 * @deprecated Use `new MongoError()` instead.
 */
MongoError.create = function(options) ***REMOVED***
  return new MongoError(options);
***REMOVED***;

/**
 * Creates a new MongoNetworkError
 * @class
 * @param ***REMOVED***Error|string|object***REMOVED*** message The error message
 * @property ***REMOVED***string***REMOVED*** message The error message
 * @property ***REMOVED***string***REMOVED*** stack The error call stack
 * @return ***REMOVED***MongoNetworkError***REMOVED*** A MongoNetworkError instance
 * @extends ***REMOVED***MongoError***REMOVED***
 */
var MongoNetworkError = function(message) ***REMOVED***
  MongoError.call(this, message);
  this.name = 'MongoNetworkError';

  // This is added as part of the transactions specification
  this.errorLabels = ['TransientTransactionError'];
***REMOVED***;
util.inherits(MongoNetworkError, MongoError);

/**
 * An error used when attempting to parse a value (like a connection string)
 *
 * @class
 * @param ***REMOVED***Error|string|object***REMOVED*** message The error message
 * @property ***REMOVED***string***REMOVED*** message The error message
 * @return ***REMOVED***MongoParseError***REMOVED*** A MongoParseError instance
 * @extends ***REMOVED***MongoError***REMOVED***
 */
const MongoParseError = function(message) ***REMOVED***
  MongoError.call(this, message);
  this.name = 'MongoParseError';
***REMOVED***;
util.inherits(MongoParseError, MongoError);

/**
 * An error signifying a timeout event
 *
 * @class
 * @param ***REMOVED***Error|string|object***REMOVED*** message The error message
 * @property ***REMOVED***string***REMOVED*** message The error message
 * @return ***REMOVED***MongoTimeoutError***REMOVED*** A MongoTimeoutError instance
 * @extends ***REMOVED***MongoError***REMOVED***
 */
const MongoTimeoutError = function(message) ***REMOVED***
  MongoError.call(this, message);
  this.name = 'MongoTimeoutError';
***REMOVED***;
util.inherits(MongoTimeoutError, MongoError);

// see: https://github.com/mongodb/specifications/blob/master/source/retryable-writes/retryable-writes.rst#terms
const RETRYABLE_ERROR_CODES = new Set([
  6, // HostUnreachable
  7, // HostNotFound
  89, // NetworkTimeout
  91, // ShutdownInProgress
  189, // PrimarySteppedDown
  9001, // SocketException
  10107, // NotMaster
  11600, // InterruptedAtShutdown
  11602, // InterruptedDueToReplStateChange
  13435, // NotMasterNoSlaveOk
  13436 // NotMasterOrSecondary
]);

function isRetryableError(error) ***REMOVED***
  if (
    RETRYABLE_ERROR_CODES.has(error.code) ||
    error instanceof MongoNetworkError ||
    error.message.match(/not master/) ||
    error.message.match(/node is recovering/)
  ) ***REMOVED***
    return true;
  ***REMOVED***

  return false;
***REMOVED***

/**
 * An error thrown when the server reports a writeConcernError
 *
 * @class
 * @param ***REMOVED***Error|string|object***REMOVED*** message The error message
 * @param ***REMOVED***object***REMOVED*** result The result document (provided if ok: 1)
 * @property ***REMOVED***string***REMOVED*** message The error message
 * @property ***REMOVED***object***REMOVED*** [result] The result document (provided if ok: 1)
 * @return ***REMOVED***MongoWriteConcernError***REMOVED*** A MongoWriteConcernError instance
 * @extends ***REMOVED***MongoError***REMOVED***
 */
const MongoWriteConcernError = function(message, result) ***REMOVED***
  MongoError.call(this, message);
  this.name = 'MongoWriteConcernError';

  if (result != null) ***REMOVED***
    this.result = result;
  ***REMOVED***
***REMOVED***;
util.inherits(MongoWriteConcernError, MongoError);

module.exports = ***REMOVED***
  MongoError,
  MongoNetworkError,
  MongoParseError,
  MongoTimeoutError,
  MongoWriteConcernError,
  mongoErrorContextSymbol,
  isRetryableError
***REMOVED***;
